### PR TITLE
Python: more additional taint steps

### PR DESCRIPTION
**Pull request by RasmusWL**
The taint steps defined here are fairly rough, and are only using syntactical matching. The idea is that this should be good enough as a first step, and then we can focus on more interesting things such as using TypeTracking :wink: 

This implementation is very generous in handing out taint. As in #4124 we don't make any attempt to limit taint propagation based on types, so since we want `<tainted dict>.get(<key>)` to propagate taint, we also propagate taint for `<any tainted obj>.get()`. I'm not too sure whether this turns out to be a problem or not.

The one thing I need want to fix is the very syntactical approach to matching library calls. This is blocked on having a proper replacement for `Value::named("json.loads").getACall()` (so we can also recognize `from json import loads; loads(tainted_string)`).

### CODE REVIEW

**Comment by RasmusWL** [comment]
should update now that https://github.com/github/codeql/pull/4161 has been merged

**Comment by RasmusWL** [comment]
@yoff no changes after merging in #4161 -- looking over the PR, it looks like there was no test of flow _from_ a PostUpdateNote. So we need some more work on PostUpdateNote, but it's not blocking this PR as I see it.

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_collections.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    ensure_tainted(
+        tainted_dict["name"],
+        tainted_dict.get("name"),
+        tainted_dict[x],
+        tainted_dict.copy(),
+    )
+
+    for v in tainted_dict.values():
+        ensure_tainted(v)
+    for k, v in tainted_dict.items(): # TODO: dict.items() currently not handled
```
Comment (yoff) [inline python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_collections.py:81]:
```suggestion
    for k, v in tainted_dict.items():
```

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    ts = TAINTED_STRING
+    import json
+
+    ensure_tainted(
+        json.dumps(ts),
+        json.loads(json.dumps(ts)),
+    )
+
+    # For Python2, need to convert to unicode for StringIO to work
+    tainted_filelike = StringIO(unicode(json.dumps(ts)))
```
Comment (yoff) [inline python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py:31]: Is `StringIO` implemented?

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    # We don't handle this case currently, since we mark `d`, `e` and `f` as tainted.
+    (a, b, c), (d, e, f) = tainted_list, no_taint_list
+    ensure_tainted(a, b, c)
+    ensure_not_tainted(d, e, f)
+
+
+def contrived_2():
+    # A contrived example. Don't know why anyone would ever actually do this.
+
+    # We currently only handle taint nested 2 levels.
```
Comment (yoff) [inline python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py:63]: This test passes, should it fail by the comment? (does the code represent 2 or 3 nested levels?)

**Review by yoff — CHANGES_REQUESTED** [review]
It is currently a little inconsistent if there are comments in the test code when the result is an expected failure.

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll [hunk 4]
```diff
@@ -123,3 +133,96 @@ predicate stringManipulation(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeT
...
+/**
+ * Holds if taint can flow from `nodeFrom` to `nodeTo` with a step related to copying.
+ */
+predicate copyStep(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeTo) {
+  exists(CallNode call | call = nodeTo.getNode() |
+    // Fully qualified: copy.copy, copy.deepcopy
+    (
+      call.getFunction().(NameNode).getId() in ["copy", "deepcopy"]
+      or
+      call.getFunction().(AttrNode).getObject("copy").(NameNode).getId() in ["copy", "deepcopy"]
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:199]: yes :facepalm:

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py (hunk 3 above)
Comment (RasmusWL) [reply python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py]: this is 3 levels `[[[`

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py (hunk 2 above)
Comment (RasmusWL) [reply python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py:31]: `StringIO` is not modeled currently.

**Comment by RasmusWL** [comment]
@yoff thanks for thorough review. I see that I forgot to update a lot of the comments for the tests based on the new implementataion :grimacing: but I fixed all the cases you pointed out now :+1:

**Review by yoff — COMMENTED** [review]
All changes look good, thanks for also adding the extra test 👍

**Review by yoff — APPROVED** [review]

**Review by tausbn — CHANGES_REQUESTED** [review]
Generally this looks good to me, but I have a few comments about how the code is structured.

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll [hunk 5]
```diff
@@ -123,3 +133,96 @@ predicate stringManipulation(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeT
...
+ * `<tainted object>.get(<arg>)` will be tainted, whether it's true or not.
+ */
+predicate containerStep(DataFlow::CfgNode nodeFrom, DataFlow::Node nodeTo) {
+  // construction by literal
+  // TODO: Not limiting the content argument here feels like a BIG hack, but we currently get nothing for free :|
+  storeStep(nodeFrom, _, nodeTo)
+  or
+  // constructor call
+  exists(CallNode call | call = nodeTo.asCfgNode() |
+    call.getFunction().(NameNode).getId() in ["list", "set", "frozenset", "dict", "defaultdict",
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:160]: Thea idea was that all these matches on `call.getFunction().(NameNode).getId()` are temporary until we have a proper replacement for `Value::named()`. It's not something I'm too concerned about fixing up. If it's something you feel very strongly about, I can give it a try :+1:

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll [hunk 6]
```diff
@@ -123,3 +133,96 @@ predicate stringManipulation(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeT
...
+    name in ["copy",
+          // general
+          "pop",
+          // dict
+          "values", "items", "get", "popitem"] and
+    call.getFunction().(AttrNode).getObject(name) = nodeFrom.asCfgNode()
+  )
+  or
+  // list.append, set.add
+  exists(CallNode call, string name |
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:182]: Personally I don't find it too problematic, so how strong is your strong urge? :blush:

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (tausbn) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:160]: Nah, I'm in the lower half of the sliding scale, given the context on this one. If we start adding a bunch more cases like these (pre-replacement), I definitely think we _should_ factor it out, if nothing else to make the replacement easier to implement.

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll (hunk 6 above)
[the remark this answers is not in the document]
Comment (tausbn) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:182]: Not _that_ strong. :slightly_smiling_face:

**Review by tausbn — APPROVED** [review]

**Pull request merged**